controlplane/src/core/bufservices/PlatformService.ts

```diff
@@ -11449,5 +11499,122 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
+            response: {
+              code: EnumStatusCode.ERR,
+              details: 'Invalid date range',
+            },
+            deliveries: [],
+            totalCount: 0,
+          };
+        }
+
+        const { deliveries, totalCount } = await orgRepo.getWebhookHistory({
+          organizationID: authContext.organizationId,
+          limit: req.pagination?.limit,
+          offset: req.pagination?.offset,
+          filterByType: req.filterByType,
+          startDate: dateRange.start,
+          endDate: dateRange.end,
+        });
+
+        return {
+          response: {
+            code: EnumStatusCode.OK,
+          },
+          deliveries,
+          totalCount,
+        };
+      });
+    },
+
+    getDeliveryDetails: (req, ctx) => {
+      let logger = getLogger(ctx, opts.logger);
+
+      return handleError<PlainMessage<GetDeliveryDetailsResponse>>(ctx, logger, async () => {
+        const authContext = await opts.authenticator.authenticate(ctx.requestHeader);
+        logger = enrichLogger(ctx, logger, authContext);
+
+        const orgRepo = new OrganizationRepository(logger, opts.db, opts.billingDefaultPlanId);
+
+        const delivery = await orgRepo.getWebhookDeliveryById(req.id, authContext.organizationId);
+        if (!delivery) {
+          return {
+            response: {
+              code: EnumStatusCode.ERR_NOT_FOUND,
+              details: `Could not find webhook delivery`,
+            },
+          };
+        }
+
+        return {
+          response: {
+            code: EnumStatusCode.OK,
+          },
+          delivery: {
+            ...delivery,
+            createdBy: delivery.user?.email || undefined,
+            isRedelivery: !!delivery.originalDeliveryId,
+            createdAt: delivery.createdAt.toISOString(),
+            requestHeaders: JSON.stringify(delivery.requestHeaders),
+            responseHeaders: delivery.responseHeaders ? JSON.stringify(delivery.responseHeaders) : undefined,
+            responseStatusCode: delivery.responseStatusCode || undefined,
+            responseErrorCode: delivery.responseErrorCode || undefined,
+            responseBody: delivery.responseBody || undefined,
+            errorMessage: delivery.errorMessage || undefined,
+          },
+        };
+      });
+    },
+
+    redeliverWebhook: (req, ctx) => {
+      let logger = getLogger(ctx, opts.logger);
+
+      return handleError<PlainMessage<RedeliverWebhookResponse>>(ctx, logger, async () => {
+        const authContext = await opts.authenticator.authenticate(ctx.requestHeader);
+        logger = enrichLogger(ctx, logger, authContext);
+
+        const orgRepo = new OrganizationRepository(logger, opts.db, opts.billingDefaultPlanId);
+        const redeliverWebhookService = new RedeliverWebhookService(opts.db, authContext.organizationId, logger);
+
+        const originalDelivery = await orgRepo.getWebhookDeliveryById(req.id, authContext.organizationId);
+        if (!originalDelivery) {
```

Comment:
Only users with write / dev access should be able to do it.

Reply:
Done